rfcs/0013-module-versioning.md

```diff
@@ -0,0 +1,161 @@
+
+  let name = "some_data"
+
+  module Stable = struct
+
+    module V1 = struct
+
+      let version = 1
+
+      module T = struct
+        type t = ... [@@deriving bin_io,sexp,yojson]
+      end
+
+      module Versioned = struct
+
+        type t =
+          { name : string
+          ; version : int
+          ; t : T.t
+          } [@@deriving bin_io,sexp,yojson]
+
+        let create t =
+          { name
+          ; version
+          ; t
+          }
+
+        let to_yojson (t : T.t) =
+          yojson_of_t @@ create t
+
+        (* similarly for binio, sexp *)
+
+      end
+    end
+  end
+end
+```
+
+Putting the name and version in the type `Versioned.t` means they'll be
+in the `sexp` and `yojson` representations, allowing any software to
+detect type version changes.
+
+Note that the definition of `Versioned` is boilerplate, and could be
+generated with an OCaml ppx.
+
+A new version of the module `V2` consists of a new version number,
+a new submodule `T`, and the same boilerplate. A ppx could
+take the version number and submodule to produce the complete
+module definition. Perhaps it would also take the list of
+serializations needed (often just one or two are needed), as
+well as other `deriving` items.
+
+It could be useful to transform data from an older version to a newer
+version. Such a transformer could be included as a ppx argument.
+
+### Serialization restricted to Stable, versioned modules
+
+Serialization should be allowed only for modules following
+this discipline, including the use of the `Stable` submodule
+containing versions.
+
+An exception can be carved out for using `sexp` serialization of data
+from modules without `Stable` and versioning. That kind of
+serialization is useful for logging and other printing from within
+Coda, and is less likely to be used with other software.
+
+### Embracing this discipline
+
+As of this writing, there are over 600 type definitions with `deriving
+bin_io` in the Coda codebase, even more with `sexp`. Only about 60
+type definitions are versioned using the `Stable.Vn` naming discipline
+(and without the explicit version information suggested here). To
+embrace this discipline fully will take some effort.
+
+We can make awareness of the existing recommendation, or the extended
+discipline suggested here a checklist item for Github pull requests.
+
+A ppx to implement the extended discipline will make it easier to comply
+with the standard.
+
+### Syntax for ppx
+
+A use of the ppx might look like:
+
+```ocaml
+[@@versioned_module
+  (`Version 1
+  ,`Type (int * string),
+  ,`Deriving [bin_io;sexp;yojson;eq]
+  )
+]
+```
+The ppx would generate the `Version` module, as above. That module would not
+contain any special code for `eq` or other non-serialization `deriving` items,
+other than to mention them in a `deriving` annotation.
+
+## Drawbacks
+
+We may not need the versioning for `sexp` and `yojson`, if these representations
+are not, in fact, used by other software.
```

Comment:
I think we probably only need the versioning for `bin_io`? The two main use cases I can think of are RPC communication with nodes running different versions of the software and being able to run a new version of the software with on-disk state created by an older version, and migrating data over to the new format.

Reply:
The versioning of modules here is independent of the RPC versioning, which is the subject of RFC 0012.

If you have a `bin_io` representation of data, then you can turn it into OCaml data, and do the migration between versions in OCaml.

Reply:
I spoke with @bkase about this, and realized a type in a versioned module could be mentioned in an RPC type. I added a para about that.